### PR TITLE
Task/generic sql aggregator

### LINKED ISSUE #1787: Extend aggregation method for every sink
Currently, only MySQL sink deals with missing attributes in batch insertions. It completes the missing fields with NULL. (WIP PostGIS and PostgreSQL)

This behaveour is the desired behaviour for every sink. At least, it has to be asessed if this is the desired behaviour for:

- CKANSink
- HDFSSink
- STHSink
- MongoSink

Moreover, if it is confirmed that this behaviour is valid for every sink, it would be interesting to make a refactor in order to include the aggregation method in the NGSISink class.

### PR DESCRIPTION
**Pull request by IvanHdzC**
Fixes #1787

### CODE REVIEW

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
+package com.telefonica.iot.cygnus.aggregation;
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: Licence header in every new file, pls.

In addition, a proper line to CHANGES_NEXT_RELEASED regarding the changes in this PR should be added.

**Comment by fgalan** [comment]
Does this PR involves changes in documentaiton (maybe some new configuration item to the sinks to which the new aggregation funcionality has been extended...)?

Probably no, just checking, any case :)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java (hunk 1 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: Fixed in [e26b297](https://github.com/telefonicaid/fiware-cygnus/pull/1798/commits/e26b29747c86742e4397bc57a6ad05d1cfe39ee8)

**Comment by IvanHdzC** [comment]
> Does this PR involves changes in documentaiton (maybe some new configuration item to the sinks to which the new aggregation funcionality has been extended...)?
> 
> Probably no, just checking, any case :)

Not really, this changes are mainly technical for optimization purposes, not functional.

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
+/**
+ * Copyright 2014-2017 Telefonica Investigación y Desarrollo, S.A.U
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]:
```suggestion
 * Copyright 2014-2020 Telefonica Investigación y Desarrollo, S.A.U
```

:D

Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: Fixed in [0d50fd2](https://github.com/telefonicaid/fiware-cygnus/pull/1798/commits/0d50fd2a1dd53ec6084568e6e2ce90bb3f64daba)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMySQLSink.java [hunk 3]
```diff
@@ -301,371 +292,85 @@ public void expirateRecords(long expirationTime) throws CygnusExpiratingError {
...
-            service = event.getServiceForNaming(enableNameMappings);
-            servicePathForData = event.getServicePathForData();
+            super.initialize(event);
             servicePathForNaming = event.getServicePathForNaming(enableGrouping, enableNameMappings);
             entityForNaming = event.getEntityForNaming(enableGrouping, enableNameMappings, enableEncoding);
             entityType = event.getEntityTypeForNaming(enableGrouping, enableNameMappings);
             attribute = event.getAttributeForNaming(enableNameMappings);
-            dbName = buildDbName(service);
-            tableName = buildTableName(servicePathForNaming, entityForNaming, entityType, attribute);
+            setDbName(buildDbName(event.getServiceForNaming(enableNameMappings)));
```
Comment (manucarrace) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIMySQLSink.java]: I'm a bit confused regarding getServiceForNaming() and getServicePathForNaming(). Should they have the same arguments? Any reason for omitting the serviceForNaming var?

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        String stringValue;
+        if (attrNativeTypes) {
+            if (value == null || value.isJsonNull()) {
+                stringValue = "NULL";
+            } else if (value.isJsonPrimitive()) {
+                if (value.getAsJsonPrimitive().isBoolean()) {
+                    stringValue = value.getAsString().toUpperCase();
+                } else if (value.getAsJsonPrimitive().isNumber()) {
+                    stringValue = value.getAsString();
+                }else {
```
Comment (AlvaroVega) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: A space expected

**Pull request merged**